great_ape_safe/ape_api/helpers/balancer/weighted_math.py

```diff
@@ -1,53 +1,244 @@
+        # // bO = balanceOut                                                                           //
+        # // bI = balanceIn              /      /            bI             \    (wI / wO) \           //
+        # // aI = amountIn    aO = bO * |  1 - | --------------------------  | ^            |          //
+        # // wI = weightIn               \      \       ( bI + aI )         /              /           //
+        # // wO = weightOut                                                                            //
+        # **********************************************************************************************/
+
+        denominator = balance_in + amount_in
+        base = divUp(balance_in, denominator)  # balanceIn.divUp(denominator);
+        exponent = divDown(weight_in, weight_out)  # weightIn.divDown(weightOut);
+        power = powUp(base, exponent)  # base.powUp(exponent);
+
+        return mulDown(balance_out, complement(power))  # balanceOut.mulDown(power.complement());
+
+    @staticmethod
+    def calc_in_given_out(balance_in: Decimal,
+                          weight_in: Decimal,
+                          balance_out: Decimal,
+                          weight_out: Decimal,
+                          amount_out: Decimal):
+
+        # /**********************************************************************************************
+        # // inGivenOut                                                                                //
+        # // aO = amount_out                                                                            //
+        # // bO = balance_out                                                                           //
+        # // bI = balance_in              /  /            bO             \    (wO / wI)      \          //
+        # // aI = amount_in    aI = bI * |  | --------------------------  | ^            - 1  |         //
+        # // wI = weight_in               \  \       ( bO - aO )         /                   /          //
+        # // wO = weight_out                                                                            //
+        # **********************************************************************************************/
+
+        base = divUp(balance_out, (balance_out - amount_out))
+        exponent = divUp(weight_out, weight_in)
+        power = powUp(base, exponent)
+        ratio = power - Decimal(1)
+        result = mulUp(balance_in, ratio)
+        return result
+
+    @staticmethod
+    def calc_bpt_out_given_exact_tokens_in(pool, balances: list, amounts_in: list):
+
+        balances = [Decimal(x) for x in balances]
+        amounts_in = [Decimal(x) for x in amounts_in]
+        normalized_weights = [Decimal(x / 1e18) for x in pool.getNormalizedWeights()]
+        bptTotalSupply = Decimal(pool.totalSupply())
+        swap_fee = Decimal(pool.getSwapFeePercentage() / 1e18)
```

Comment:
im wondering if it wouldnt be better to do these type of conversions using a helper func in the balancer class? this way weighted_math.py can remain untouched and used completely raw from source. dunno.